### PR TITLE
added the addQuestionToFormTemplate in the admin service

### PR DESCRIPTION
**Pull request by dhmi19**
## Notion ticket link

[Form Management: Add question to form template](https://www.notion.so/uwblueprintexecs/Task-Board-F22-9e03bf9f7c4343f09a8a5c16dc48a12a?p=872084408d3543a3ad120dcc9808a9a1&pm=c)

## Implementation description
* just your standard endpoint changes tbh
* decided to make the body of the form `formQuestion: {...}` instead of just the fields. This was to leverage the existing `validateFormQuestion()` function more directly. 
* Decided to implement this function in the adminService as I think it does belong there. Some of the existing work for formquestions seems to be in the `campService` but I think it's better suited here. (For camp creation, I think we should add new functions like `addQuestionsToCamp` like the one I added in this PR)

## Steps to test
1. Make a PATCH request to `http://localhost:5000/admin/formTemplate/formQuestion/` (you might need to disable the `isAuthorizedByRole()` middleware or figure out how to add auth to POSTMAN requests)
2. Use the following body: 
```
{
    "formQuestion" : {
        "type" : "Text",
        "category": "PersonalInfo",
        "question": "What previous photography experience do you have?",
        "required" : false
    }
}
```
3. Log on to the mongodb console and see that the question id was APPENDED to the list of questions in the template and that the question was created in the `formQuestion` collection
4. The response should have the newly created formQuestion

## What should reviewers focus on?
* Use of transactions ok? 

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [ ] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

**Comment by dhmi19** [comment]
### TEST CASES

## Case 1: Happy path
body : 
```
{
    "formQuestion" : {
        "type" : "Text",
        "category": "PersonalInfo",
        "question": "What previous photography experience do you have?",
        "required" : false
    }
}
```
response:
```
{
    "type": "Text",
    "question": "What previous photography experience do you have?",
    "required": false,
    "options": [],
    "id": "6358ae8d06d07d99bcf5d383",
    "category": "PersonalInfo"
}
```

##Case 2: Unauthorized
Same body as above but this time had the middleware uncommented and no authorization on the POSTMAN request
response:
```
{
    "error": "You are not authorized to make this request."
}
```

##Case 3: Bad body (missing the field `required`) 
body:
```
{
    "formQuestion" : {
        "type" : "Text",
        "category": "PersonalInfo",
        "question": "What previous photography experience do you have?"
    }
}
```
response: `Invalid format provided for form question`

File: backend/typescript/services/implementations/adminService.ts [hunk 1]
```diff
@@ -108,6 +114,52 @@ class AdminService implements IAdminService {
...
+        category: newFormQuestion.category,
+      };
+    } catch (error: unknown) {
+      Logger.error(
+        `Failed to add form question to form template. Reason = ${getErrorMessage(
+          error,
+        )}`,
+      );
+      await session.abortTransaction();
+      await session.endSession();
```
Comment (e-wai) [inline backend/typescript/services/implementations/adminService.ts:159]: should this be moved to a `finally {}`? Even if no error, we should end session

File: backend/typescript/services/implementations/adminService.ts [hunk 2]
```diff
@@ -108,6 +114,52 @@ class AdminService implements IAdminService {
...
+    formQuestion: CreateFormQuestionDTO,
+  ): Promise<FormQuestionDTO> {
+    const session = await mongoose.startSession();
+    session.startTransaction();
+
+    try {
+      // Get the current formTemplate and its list of questions
+      const formTemplate = await MgFormTemplate.findOne();
+      if (!formTemplate)
+        throw new Error("Failed to retrieve the form template");
```
Comment (e-wai) [inline backend/typescript/services/implementations/adminService.ts:128]: nit: brackets

File: backend/typescript/services/implementations/adminService.ts [hunk 3]
```diff
@@ -108,6 +114,52 @@ class AdminService implements IAdminService {
...
+      const newFormQuestion = await MgFormQuestion.create({
+        type: formQuestion.type,
+        question: formQuestion.question,
+        required: formQuestion.required,
+        description: formQuestion.description,
+        options: formQuestion.options,
+        category: formQuestion.category,
+      });
+
+      formTemplate.formQuestions.push(newFormQuestion._id);
```
Comment (e-wai) [inline backend/typescript/services/implementations/adminService.ts:139]: I've completely forgotten, what's the difference between `_id` and `id`?

**Comment by e-wai** [comment]
> Make a PATCH request to http://localhost:5000/admin/formTemplate/formQuestion/ (you might need to disable the isAuthorizedByRole() middleware or figure out how to add auth to POSTMAN requests)

I think this restriction is figured out by your auth token so should be good in POSTMAN

File: backend/typescript/middlewares/validators/adminValidators.ts [hunk 4]
```diff
@@ -49,3 +49,14 @@ export const formTemplateUpdateValidator = async (
...
   return next();
 };
+
+export const formTemplateAddQuestionValidator = async (
+  req: Request,
+  res: Response,
+  next: NextFunction,
+) => {
+  if (!req.body.formQuestion || !validateFormQuestion(req.body.formQuestion)) {
+    return res.status(400).send("Invalid format provided for form question");
```
Comment (e-wai) [inline backend/typescript/middlewares/validators/adminValidators.ts:59]: I think it's worth specifying which field we're missing?

**Review by e-wai — CHANGES_REQUESTED** [review]
Transactions 😍 so much cleaner, also tysm for the documentation on the testing cases, works like a charm

> Decided to implement this function in the adminService as I think it does belong there. Some of the existing work for formquestions seems to be in the campService but I think it's better suited here. (For camp creation, I think we should add new functions like addQuestionsToCamp like the one I added in this PR)

yep agree

- Should this be a POST endpoint? It always creates new resources
- I also didn't realize none of our endpoints are protected... they very much should be

File: backend/typescript/rest/adminRoutes.ts [hunk 5]
```diff
@@ -54,4 +57,25 @@ adminRouter.get("/formTemplate", async (req, res) => {
   }
 });
 
+adminRouter.patch(
+  "/formTemplate/formQuestion",
+  formTemplateAddQuestionValidator,
+  // isAuthorizedByRole(new Set(["Admin"])),
```
Comment (e-wai) [inline backend/typescript/rest/adminRoutes.ts:63]: Could we uncomment this (to test, I can just change my permissions)

File: backend/typescript/services/implementations/adminService.ts (hunk 3 above)
Comment (dhmi19) [reply backend/typescript/services/implementations/adminService.ts:139]: the id is the string representation and _id is the actual OjectId type

**Comment by dhmi19** [comment]
> Transactions 😍 so much cleaner, also tysm for the documentation on the testing cases, works like a charm
> 
> > Decided to implement this function in the adminService as I think it does belong there. Some of the existing work for formquestions seems to be in the campService but I think it's better suited here. (For camp creation, I think we should add new functions like addQuestionsToCamp like the one I added in this PR)
> 
> yep agree
> 
> * Should this be a POST endpoint? It always creates new resources
> * I also didn't realize none of our endpoints are protected... they very much should be

I was considering POST but we are technically updating the form template. I don't think we can ever create a form question by itself (the creation will always either be part of updating the template or the camp's list of form questions)

File: backend/typescript/services/implementations/adminService.ts (hunk 1 above)
Comment (dhmi19) [reply backend/typescript/services/implementations/adminService.ts:161]: good callout

**Review by e-wai — APPROVED** [review]
LGTM, just linter outstanding

File: backend/typescript/services/implementations/adminService.ts [hunk 6]
```diff
@@ -108,6 +114,55 @@ class AdminService implements IAdminService {
     }
     return form;
   }
+
+  async addQuestionToTemplate(
+    formQuestion: CreateFormQuestionDTO,
+  ): Promise<FormQuestionDTO> {
+    const session = await mongoose.startSession();
+    session.startTransaction();
```
Comment (rchang0501) [inline backend/typescript/services/implementations/adminService.ts:122]: ah finally understand what a transaction is - cool!

File: backend/typescript/services/interfaces/adminService.ts [hunk 7]
```diff
@@ -8,24 +13,36 @@ interface IAdminService {
...
    * Get the form.
    * @returns A FormTemplateDTO that contains the waiver.
    * @throws Error if form template retrieval fails.
    */
   getFormTemplate(): Promise<FormTemplateDTO>;
+
+  /**
+   * Add a question to the form template
+   * @returns a FormQuestionDTO for the new form
+   * @throws Error if adding the form question fails
```
Comment (rchang0501) [inline backend/typescript/services/interfaces/adminService.ts:41]: should we also define params? I know we use them in the camper service interface, how come we don't use them here?

**Review by rchang0501 — APPROVED** [review]
lgtm, just small comment for my own curiosity

**Comment by dhmi19** [comment]
TODO: forgot to include {session} in my mongoose db calls

File: backend/typescript/services/interfaces/adminService.ts (hunk 7 above)
Comment (dhmi19) [reply backend/typescript/services/interfaces/adminService.ts:41]: good catch. I forgot about those

**Comment by e-wai** [comment]
> TODO: forgot to include {session} in my mongoose db calls

Like as an import? How did it work without the import?

File: backend/typescript/middlewares/validators/adminValidators.ts (hunk 4 above)
Comment (e-wai) [reply backend/typescript/middlewares/validators/adminValidators.ts:59]: Yeah, fair enough

**Pull request merged**